Display player statistics

BlackMorzan opened the pull request:
Enabling user to see how many times per question he give incorrect answer and when he answered correctly 

## Description

New tab in Menage Account. Allows user to see when did he answered, and how many times did he answered wrong.
Shows only questions that have been answered.

## Where should the reviewer start?


## Motivation and context
closes #388

## How has this been tested?

Tested for:
- no tests
- no wrong and correct answers
- no correct answers
- no wrong answers 
- some wrong and correct answers and 

## Screenshots (if appropriate)

![Viz](https://user-images.githubusercontent.com/44707515/99822283-4f70bb00-2b53-11eb-9674-02c644e4d265.png)


![Pytania](https://user-images.githubusercontent.com/44707515/98701989-2a719080-237a-11eb-8a47-e0d082f3b017.png)

![brak_pytan](https://user-images.githubusercontent.com/44707515/98701957-22b1ec00-237a-11eb-88e3-93f248d9ce7a.png)

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.TestResultService/TestResultRepository.cs:
```diff
@@ -198,5 +198,34 @@ public ApplicationUser GetUserById(string id)
+        {
+            return _dbContext
+                .UserTestWrongAnswers
+                .Where(a => a.TestId == testID)
```
use `.Where(a=> a.TestId == testId & a.UserId == userId)`

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.TestResultService/TestResultRepository.cs:
```diff
@@ -198,5 +198,34 @@ public ApplicationUser GetUserById(string id)
+         * 
+         *  [out] Count of submitted wrong answers
+         */
+        public int GetWrongAnswersToTest(string userId, int testID) // get count of wrong answers
```
I suggest renaming it to _GetUserTestWrongAnswerCount_

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.TestResultService/TestResultRepository.cs:
```diff
@@ -198,5 +198,34 @@ public ApplicationUser GetUserById(string id)
+        {
+            return _dbContext
+                .UserTestCorrectAnswers
+                .Where(a => a.UserId == userId)
```
as above -  join _Where()_ clause

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.TestResultService/TestResultRepository.cs:
```diff
@@ -198,5 +198,34 @@ public ApplicationUser GetUserById(string id)
+                .Where(a => a.UserId == userId)
+                .Where(a => a.TestId == testID)
+                .Select(a => a.AnsweringTime) 
+                .First(); // I assume there is only one row
```
it can be _Single()_, but _First()_ is ok, too

BlackMorzan replied:
Agree. Single makes more sense here

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.TestResultService/TestResultRepository.cs:
```diff
@@ -198,5 +198,34 @@ public ApplicationUser GetUserById(string id)
+         * 
+         *  [out] Date of submission of corect answer
+         */
+        public DateTime GetCorrectAnswerDate(string userId, int testID) // get DateTime of correct answer
```
I suggest renaming it to _GetUserTestCorrectAnswer_ - remember, that it can be null!!

BlackMorzan replied:
I'd say GetUserTestCorrectAnswer suggest that method returns whole CorrectAnswer instead of DateTime of CorrectAnswer. How about GetUserTestCorrectAnswerDate?

BlackMorzan commented:
in last commits I:
1. Replaced ITestStatisticsService to namespace .../Services.Interfaces 
2. Added base for StatisticsController controller
3. Added base for StatisticsController unit tests

My next goal is to make base for integration tests, but I'm unsure how to start.
I will also try to make smaller commits in future when I have covered basics.

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/UnitTests/StatisticsServiceTests.cs:
```diff
@@ -0,0 +1,65 @@
+{
+    public class StatisticsServiceTests
+    {
+        private readonly StatisticsService _controller;
```
service or controller? :)

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/UnitTests/StatisticsServiceTests.cs:
```diff
@@ -0,0 +1,65 @@
+            this._testStatisticsMock = new Mock<ITestStatisticsService>();
+
+            // might be useless
+            this._testUserTestCorrectAnswerMock = new Mock<UserTestCorrectAnswer>();
```
yes, useless - you define for example mock of service

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/UnitTests/StatisticsServiceTests.cs:
```diff
@@ -0,0 +1,65 @@
+        {
+            int a = this._controller.TmpStatisticsImpementation("1", 2);
+
+            Assert.True(a == 3, "Temoporary test - wrong!");
```
please use 3A rule :)

BlackMorzan replied:
I will use A3 rule when tests will be meaningfull :)

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/Controllers/ManageController.cs:
```diff
@@ -272,6 +274,34 @@ public async Task<IActionResult> SetPassword(SetPasswordViewModel model)
+            string messsageToUser = this._testStatisticsService.GetCorrectAnswerDateTime(user.Email, 1).ToString(CultureInfo.InvariantCulture);
+            if (messsageToUser == null)
+            {
+                messsageToUser = "NoAnswer";
```
remember to use resources
I suggest more descriptive text for example "Nie udzieliłeś jeszcze poprawnej odpowiedzi na żadne pytanie"

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/Resources/ViewsMessages.resx:
```diff
@@ -543,6 +546,9 @@
   <data name="StartTime" xml:space="preserve">
     <value>Czas rozpoczęcia:</value>
   </data>
+  <data name="StatistisTab" xml:space="preserve">
```
a typo - should be Statistics

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Services/StatisticsService.cs:
```diff
@@ -0,0 +1,36 @@
+    {
+        private readonly ITestStatisticsService _statisticsService;
+
+        public StatisticsService(ITestStatisticsService statisticsService)
```
please simplify - Statistics or TestStatistics?

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Services/StatisticsService.cs:
```diff
@@ -0,0 +1,36 @@
+            return this._statisticsService.GetUserTestCorrectAnswerDate(playerID, testID);
+        }
+
+        public int TmpStatisticsImpementation(string playerID, int testID)
```
TODO: remove

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Services/StatisticsService.cs:
```diff
@@ -0,0 +1,36 @@
+            this._statisticsService = statisticsService;
+        }
+
+        public int GetWrongAnswerCount(string playerID, int testID)
```
please align to conversion: playerId, not playerID

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.TestResultService/TestResultRepository.cs:
```diff
@@ -198,5 +198,32 @@ public ApplicationUser GetUserById(string id)
+        {
+            return _dbContext
+                .UserTestWrongAnswers
+                .Where(a => a.TestId == testID && a.UserId == userId)
```
you can use Count(a => a.TestId == testID && a.UserId == userId)

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.TestResultService/TestResultRepository.cs:
```diff
@@ -198,5 +198,32 @@ public ApplicationUser GetUserById(string id)
+                .UserTestCorrectAnswers
+                .Where(a => a.TestId == testID && a.UserId == userId)
+                .Select(a => a.AnsweringTime) 
+                .Single(); // I assume there is only one row
```
please remove comment
it will fail if there is no correct answer - I suggesting using SingleOrDefault()

BlackMorzan replied:
I resolved it different, but this seems better

WTobor replied:
I don't think so :(

BlackMorzan replied:
There is something going on that I don't understand. Those methods have been moved to .../Repositiory/StatisticsRepository in commit "Cleanup in Repository and Services". Since then I didn't touch anything in TestResoultSevice.

BlackMorzan commented:
I can see that models like "TestResultDto" are used to send data from service to controller. Should I also make another model like this?

BlackMorzan commented:
I have now 3 questions:
- I need to test this feature, but for that I need to fill some data. I there some preferred way to do that? (I will make some Unit test in next 2 commits)
- I need to change how I find TestId (I assumed that questions have Id from 0 to 24), but I'm unsure how it works on our DataBase.
- I also assumed that UserTestCorrectAnswer and UserTestWrongAnswer are generated as soon as player gives answer. Is that true?

WTobor commented:
1. Look here https://github.com/DevAdventCalendar/DevAdventCalendar/blob/develop/src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/IntegrationTests/TestServiceTests.cs
2. Id from 1. But I would suggest searching by test number (it should be unique).
3. I don't understand. Resolve conflicts - keep your changes with changes made on develop (in PR should be visible only your changes)
4. Si

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Services/StatisticsService.cs:
```diff
@@ -0,0 +1,38 @@
+        }
+
+        // TODO: unfix max testId
+        public List<DisplayStatisticsDto> FillResultsWithTestStats(string userId)
```
TODO: remove

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/Views/Manage/DisplayStatistics.cshtml:
```diff
@@ -16,19 +16,19 @@
     </thead>
     <tbody>
-        @foreach (var i in Model)
+        @foreach (var curresntStat in Model)
```
a typo

WTobor commented:
@BlackMorzan any update here? :) There has been no notification for a week :(

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/UnitTests/StatisticsServiceTests.cs:
```diff
@@ -0,0 +1,137 @@
+            // ARRANGE
+            string userId = "c611530e-bebd-41a9-ace2-951550edbfa0";
+
+            // I assume it is wrong
```
why you are assuming it is wrong?

BlackMorzan replied:
bad wording - I assumed that the way I created those tests isn't right. Looks like there is too many Mocks

WTobor replied:
I don't think so. So maybe just remove the comment and I will close PR

WTobor replied:
@BlackMorzan please resolve conflicts

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/UnitTests/StatisticsServiceTests.cs:
```diff
@@ -0,0 +1,137 @@
+        {
+            // ARRANGE
+            string userId = "c611530e-bebd-41a9-ace2-951550edbfa0";
+            int twoCorrectAnswers = 2;
```
it should be something like `testNumberWithTwoCorrectAnswers`

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/UnitTests/StatisticsServiceTests.cs:
```diff
@@ -0,0 +1,137 @@
+        {
+            // ARRANGE
+            string userId = "c611530e-bebd-41a9-ace2-951550edbfa0";
+            int twoCorrectAnswersAreHigher = 2;
```
it should be something like `testNumber`

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/UnitTests/StatisticsServiceTests.cs:
```diff
@@ -0,0 +1,137 @@
+        {
+            // ARRANGE
+            string userId = "c611530e-bebd-41a9-ace2-951550edbfa0";
+            int twoWrongAnswers = 2;
```
it should be something like `testNumber`

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/UnitTests/StatisticsServiceTests.cs:
```diff
@@ -0,0 +1,137 @@
+        {
+            // ARRANGE
+            string userId = "c611530e-bebd-41a9-ace2-951550edbfa0";
+            int noTests = 0;
```
it should be something like `testNumber`

WTobor commented on src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/Controllers/ManageController.cs:
```diff
@@ -24,20 +26,23 @@ public class ManageController : Controller
             this._logger = logger ?? throw new ArgumentNullException(nameof(logger));
             this._emailNotificationService = emailNotificationService ?? throw new ArgumentNullException(nameof(emailNotificationService));
             this._googleCalendarService = googleCalendarService ?? throw new ArgumentNullException(nameof(googleCalendarService));
+            this._statisticsService = statisticsService ?? throw new ArgumentNullException(nameof(statisticsService)); // Temporary
```
remove Temporary comment

WTobor commented:
@BlackMorzan remember about conflicts ;)
